sys/include/riot_hdr.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2018 Kaspar Schleiser <kaspar@schleiser.de>
+ *                    Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_riot_hdr RIOT header helpers and tools
+ * @ingroup     sys
+ * @{
+ *
+ * The header contains
+ *
+ * - "RIOT" as magic number
+ * - the application version
+ * - the address where to find the RIOT firmware
+ * - the checksum of the three previous fields
+ *
+ * @file
+ * @brief       RIOT firmware header and tools
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Francisco Acosta <francisco.acosta@inria.fr>
+ *
+ * @}
+ */
+
+#ifndef RIOT_HDR_H
+#define RIOT_HDR_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+
+/**
+ * @brief  Magic number for riot_hdr
+ *
+ */
+#define RIOT_HDR_MAGIC         0x544f4952 /* "RIOT" */
+
+/**
+ * @brief Structure to store image header - All members are little endian
+ * @{
+ */
+typedef struct {
+    uint32_t magic_number;         /**< header magic_number (always "RIOT")    */
+    uint32_t version;              /**< Integer representing firmware version  */
+    uint32_t start_addr;           /**< Start address in flash                 */
+    uint32_t chksum;               /**< checksum of riot_hdr                   */
+} riot_hdr_t;
```

Comment:
@kYc0o @cladmi as tentatively discussed with @cladmi a few days ago:
would it be simple to put the chksum in front of the struct, and add in the struct the size `image_size`
and then compute the checksum:
- starting from where the hdr + offset of sizeof(chksum)
- stopping at 256 + image_size ?

then `riot_hdr_checksum` could be easy to modify so that it checks the Flash of the image too (and not only the Flash of the hdr).

Reply:
This would lead to increased boot-time (also from deep sleep wakeups).
Also, checking the whole image on boot makes it impossible to use part of the slot for dynamic data, e.g., a file system or configuration area.

Also I don't see any practical advantage in having the bootloader verify the flash over and over again.

The safety (not security) concept for flashing a slot is as follows:

(0. crypto-verify header & image)
1. erase header block (so it's checksum is invalid)
2. write image
3. verify image hash
4. write header block (including it's own hash)

If followed, that procedure makes sure that the image after the header is the one that was supposed to be flashed, should the header be valid (have a correct checksum).

So for the price of having do wait a couple of miliseconds *at each each wakeup*, we'd check whether the flash got corrupted *after* it was already flashed & verified once.

So what does the bootloader do if the newest slot has a broken checksum? With the current logic, it would boot an old (and possibly insecure) image, if one is available. The sane thing would be to not boot at all in that case, leading to a dead device. Is that desirable for all applications? Anyhow, the problem scope just got a lot larger.

I was very happy to have arrived at a very minimal bootloader. I suggest we keep it that way. If any image want's to check its own integrity, let the image check itself.

Reply:
While I don't see any problems with putting the checksum as first element of the struct, I agree with @kaspar030 here. The function of the checksum is to be able to distinguish a filled slot from an empty (or corrupt `riot_hdr_t`). Extending the function of the checksum leads to a more complex design at almost no extra guarantees.

If we want to verify an image, I think it would be better to add a separate (sha-based) hash to the first flash page, somewhere after the metadata and have a separate bootloader application that is able to check the image integrity.

Reply:
I agree with both @kaspar030 and @bergzand, this minimal header is just for booting purposes, and afaik we don't aim to provide security or even safety at this level, only a way to recognise RIOT images which meet the requirements of a consistent header.

Reply:
@kYc0o @kaspar030 @bergzand ok so let's take on this discussion in follow-up PRs / enhancements.

Reply:
I think it is also a good idea to keep the magic number as the first element of the struct. This way we never get into trouble if there are different type of headers marked with different magic numbers

Reply:
Good catch indeed.